packages/cozy-harvest-lib/src/helpers/konnectorBlock.js

```diff
@@ -0,0 +1,227 @@
+      sourceAccount
+    )
+
+    if (!accountConnected) {
+      return formatKonnector({
+        client,
+        konnector: konnector.attributes,
+        universalLink: {
+          slug: 'home',
+          nativePath: `connected/${slug}/new`
+        },
+        iconStatus: 'disabled',
+        message: {
+          text: t('konnectorBlock.disconnected')
+        }
+      })
+    }
+
+    const inMaintenance = await konnectorBlock.isInMaintenance(client, slug)
+
+    if (inMaintenance) {
+      return formatKonnector({
+        client,
+        konnector: konnector.attributes,
+        universalLink: {
+          slug: 'home',
+          nativePath: `connected/${slug}`
+        },
+        iconStatus: 'disabled',
+        message: {
+          text: t('konnectorBlock.inMaintenance')
+        }
+      })
+    }
+
+    const { error } = await konnectorBlock.isInError({
+      client,
+      slug,
+      sourceAccount
+    })
+
+    if (error) {
+      const color =
+        error.isActionable || hasNewVersionAvailable(konnector)
+          ? 'error'
+          : 'textSecondary'
+
+      return formatKonnector({
+        client,
+        konnector: konnector.attributes,
+        universalLink: {
+          slug: 'home',
+          nativePath: `connected/${slug}`
+        },
+        iconStatus: 'disabled',
+        message: {
+          text: t(`error.job.${error.message}.title`),
+          color
+        }
+      })
+    }
+
+    if (hasNewVersionAvailable(konnector)) {
+      return formatKonnector({
+        client,
+        konnector: konnector.attributes,
+        universalLink: {
+          slug: 'home',
+          nativePath: `connected/${slug}`
+        },
+        message: {
+          text: t('konnectorBlock.hasNewVersionAvailable'),
+          color: 'primary'
+        }
+      })
+    }
+
+    return formatKonnector({
+      client,
+      konnector: konnector.attributes,
+      universalLink: {
+        slug: 'home',
+        nativePath: `connected/${slug}/accounts/${sourceAccount}`
+      }
+    })
+  } catch (error) {
+    // The konnector can be uninstalled and returned 404
+    if (error.status === 404) {
+      try {
+        const konnector = await konnectorBlock.getKonnector(
+          client,
+          buildKonnectorQuery(slug, 'registry')
+        )
+
+        return formatKonnector({
+          client,
+          konnector: konnector.attributes,
+          universalLink: {
+            slug: 'store',
+            nativePath: `discover/${slug}`
+          },
+          iconStatus: 'disabled'
+        })
+      } catch (error) {
+        return {
+          fatalError: t('konnectorBlock.fatalError', {
+            name: 'support@cozycloud.cc',
+            link: 'mailto:support@cozycloud.cc'
+          })
+        }
+      }
+    }
+
+    return {
+      fatalError: t('konnectorBlock.fatalError', {
+        name: 'support@cozycloud.cc',
+        link: 'mailto:support@cozycloud.cc'
+      })
+    }
+  }
+}
```

Comment:
I'm afraid the logic of this function is hard to understand. A lot of functions could work on objects coming from the state, instead of themselves being aware of the query and dealing the client state.

It would remove a lot of logic from isInMaintenance, isInError etc...

I've made a short example to hopefully better describe the idea:

```
const useKonnectorData = () => {
  // only 1 functions deals with fetching the data via cozy-client
  const triggerQDef = buildKonnectorTriggerQuery(konnector)
  const triggerResp = useQuery(triggerQDef.query, triggerQDef)
  const accountQDef = buildKonnectorAccountQuery(konnector)
  const accountResp = useQuery(accountQDef.query, accountQDef)
  const konnectorQDef = buildKonnectorQuery(konnector)
  const konnectorResp = useQuery(konnectorQDef.query, konnectorQDef)

  // recreate a status so that consumers of the hook can only render
  // something when fetchStatus is not loading
  const isLoading = any([triggerResp, accountResp, konnectorResp], x => x.fetchStatus === 'loading')
  const errors = map([triggerResp, accountResp, konnectorResp], x => x.error)

  // other functions deal directly with cached objects
  const isInMaintenance = getMaintenanceStatus(konnector)

  return {
    account: accountResp.data,
    trigger: triggerResp.data,
    konnector: konnectorResp.data,
    isInMaintenance,
    isLoading
  }
}
```

Reply:
Les 130 lignes sont dues en grande partie à `formatKonnector` qui doit renvoyer un résultat différent selon les cas. Chose manquante dans l'exemple que tu proposes. Donc je suis d'accord qu'on doit pouvoir faire quelque chose de moins verbeux, mais je pense qu'on n'aura pas une énorme différence.

De plus ici on exécute les requêtes en parallèle, chose que j'ai souhaité évité... peut-être inutilement ? Je suis preneur d'un petit échange à ce sujet : quid du nombre de requêtes vs complexité du code ?

Enfin il ne faut pas oublier tous les cas et conditions qu'on souhaite tester, je ne sais pas si un tel code permettrait la même granularité.

En l'état je t'avoue que je me vois mal tout éclater... le choix initial d'éviter le useQuery et de partir sur des requêtes en séries à peut être était le mauvais, je suis preneur d'un retour là-dessus pour le code futur, mais pour celui-ci je pense que les tests assurent le fait qu'on a quelque chose qui répond au besoin, et qui reste assez robuste malgré tout.